From 1f18baaaef921e9b0aa514f0e3a1ff344752b3bc Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Fri, 17 Jan 2025 13:53:50 +0800
Subject: [PATCH 145/236] mtk: hostapd: add Adv-TTLM helper functions

Add 2 Adv-TTLM helper functions
1. hostapd_is_attlm_active(): check whether the hapd has a setup a
TTLM other than the default one.
2. hostapd_same_attlm(): compare the TTLMs of 2 hapds.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 hostapd/ctrl_iface.c    |  5 +++--
 src/ap/hostapd.c        |  9 +++++++++
 src/ap/hostapd.h        |  1 +
 src/ap/ieee802_11.c     |  2 +-
 src/ap/ieee802_11_eht.c | 13 +++----------
 5 files changed, 17 insertions(+), 13 deletions(-)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index 00f8fc0ae..0912ca6fe 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -4101,12 +4101,13 @@ static int hostapd_ctrl_iface_set_attlm(struct hostapd_data *hapd, char *cmd,
 	if (!hapd->conf->mld_ap || !hapd->mld)
 		return -1;
 
-	attlm = &hapd->mld->new_attlm;
-	if (attlm->valid) {
+	if (hostapd_is_attlm_active(hapd)) {
 		wpa_printf(MSG_ERROR, "Busy: A-TTLM is on-going");
 		return -1;
 	}
 
+	attlm = &hapd->mld->new_attlm;
+
 	for_each_mld_link(h, hapd)
 		valid_links |= BIT(h->mld_link_id);
 
diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 20dd93b53..290f40c98 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -490,6 +490,15 @@ void hostapd_link_remove_cb(struct hostapd_data *hapd, u16 removed_links)
 }
 
 
+bool hostapd_is_attlm_active(struct hostapd_data *hapd)
+{
+	if (!hostapd_is_mld_ap(hapd) || !hapd->mld)
+		return false;
+
+	return hapd->mld->new_attlm.valid;
+}
+
+
 int hostapd_mld_set_attlm(struct hostapd_data *hapd)
 {
 	if (!hapd->drv_priv)
diff --git a/src/ap/hostapd.h b/src/ap/hostapd.h
index e3856e2be..e1eb62178 100644
--- a/src/ap/hostapd.h
+++ b/src/ap/hostapd.h
@@ -1044,6 +1044,7 @@ int hostapd_fill_cca_settings(struct hostapd_data *hapd,
 
 #ifdef CONFIG_IEEE80211BE
 
+bool hostapd_is_attlm_active(struct hostapd_data *hapd);
 void hostapd_event_attlm(struct hostapd_data *hapd, struct attlm_event *attlm_event);
 void hostapd_event_tsf_offset(struct hostapd_data *hapd, s64 *tsf_offset);
 bool hostapd_mld_is_first_bss(struct hostapd_data *hapd);
diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 1ff6eafa9..ed80d1cee 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -6762,7 +6762,7 @@ int ieee802_11_mgmt(struct hostapd_data *hapd, const u8 *buf, size_t len,
 		return 1;
 	}
 
-	if (hapd->conf->mld_ap && hapd->mld->new_attlm.valid &&
+	if (hostapd_is_attlm_active(hapd) &&
 	    !hapd->mld->new_attlm.switch_time_tsf_tu &&
 	    (hapd->mld->new_attlm.disabled_links & BIT(hapd->mld_link_id))) {
 		wpa_printf(MSG_DEBUG, "MGMT: Drop the frame - Disabled link");
diff --git a/src/ap/ieee802_11_eht.c b/src/ap/ieee802_11_eht.c
index 3e82f37b7..d4cf732b3 100644
--- a/src/ap/ieee802_11_eht.c
+++ b/src/ap/ieee802_11_eht.c
@@ -912,14 +912,9 @@ static u8 * hostapd_eid_eht_reconf_ml(struct hostapd_data *hapd, u8 *eid)
 
 size_t hostapd_eid_eht_attlm_len(struct hostapd_data * hapd)
 {
-	struct attlm_settings *attlm;
 	size_t len;
 
-	if (!hapd->conf->mld_ap)
-		return 0;
-
-	attlm = &hapd->mld->new_attlm;
-	if (!attlm || !attlm->valid)
+	if (!hostapd_is_attlm_active(hapd))
 		return 0;
 
 	/* Element ID: 1 octet
@@ -931,7 +926,7 @@ size_t hostapd_eid_eht_attlm_len(struct hostapd_data * hapd)
 	 * Link Mapping (size of 2) for All 8 TIDs: 2 * 8 octets
 	 */
 	len = 3 + 2 + 3 + 2 * 8;
-	if (attlm->switch_time_tsf_tu != 0)
+	if (hapd->mld->new_attlm.switch_time_tsf_tu != 0)
 		len += 2;
 
 	return len;
@@ -947,12 +942,10 @@ u8 * hostapd_eid_eht_attlm(struct hostapd_data *hapd, u8 *eid)
 	u8 *pos = eid;
 	u16 enabled_links;
 
-	if (!hapd->conf->mld_ap)
+	if (!hostapd_is_attlm_active(hapd))
 		return eid;
 
 	attlm = &hapd->mld->new_attlm;
-	if (!attlm || !attlm->valid)
-		return eid;
 
 	/* The length will be set at the end */
 	*pos++ = WLAN_EID_EXTENSION;
-- 
2.45.2

